smt/surrogate_models/neural_net/demo.py

```diff
@@ -0,0 +1,180 @@
+
+    # Domain
+    lb = -1.  # minimum bound (same for all dimensions)
+    ub = 1.5  # maximum bound (same for all dimensions)
+
+    # Design of experiment (full factorial)
+    n_x = 2  # number of dimensions
+    n_y = 1  # number of responses
+    L = 12  # number of levels per dimension
+    m = L ** n_x  # number of training examples that will be generated
+
+    if random:
+        doe = np.random.rand(m, n_x)
+    else:
+        levels = [L] * n_x
+        doe = fullfact(levels)
+        doe = (doe - 0.) / (L - 1.)  # values normalized such that 0 < doe < 1
+
+    assert (doe.shape == (m, n_x))
+
+    # Apply bounds
+    X = lb + (ub - lb) * doe
+
+    # Evaluate response
+    Y = f(X).reshape((m, 1))
+
+    # Evaluate partials
+    J = np.zeros((m, n_x, n_y))
+    for j in range(0, n_x):
+        J[:, j, :] = df(X, j).reshape((m, 1))
+
+    return X.T, Y.T, J.T
+
+
+def contour_plot(model, title='GENN'):
+    """Make contour plots of 2D Rastrigin function and compare to Neural Net prediction"""
+
+    X_train, _, _ = model.training_data
+
+    # Domain
+    lb = -1.
+    ub = 1.5
+    m = 100
+    x1 = np.linspace(lb, ub, m)
+    x2 = np.linspace(lb, ub, m)
+    X1, X2 = np.meshgrid(x1, x2)
+
+    # True response
+    pi = np.pi
+    Y_true = np.power(X1, 2) - 10 * np.cos(2 * pi * X1) + 10 + np.power(X2, 2) - 10 * np.cos(2 * pi * X2) + 10
+
+    # Predicted response
+    Y_pred = np.zeros((m, m))
+    for i in range(0, m):
+        for j in range(0, m):
+            Y_pred[i, j] = model.evaluate(np.array([X1[i, j], X2[i, j]]).reshape(2, 1))
+
+    # Prepare to plot
+    fig = plt.figure(figsize=(6, 3))
+    spec = gridspec.GridSpec(ncols=2, nrows=1, wspace=0)
+
+    # Plot Truth model
+    ax1 = fig.add_subplot(spec[0, 0])
+    ax1.contour(X1, X2, Y_true, 20, cmap='RdGy')
+    anno_opts = dict(xy=(0.5, 1.075), xycoords='axes fraction', va='center', ha='center')
+    ax1.annotate('True', **anno_opts)
+    anno_opts = dict(xy=(-0.075, 0.5), xycoords='axes fraction', va='center', ha='center')
+    ax1.annotate('X2', **anno_opts)
+    anno_opts = dict(xy=(0.5, -0.05), xycoords='axes fraction', va='center', ha='center')
+    ax1.annotate('X1', **anno_opts)
+    ax1.set_xticks([])
+    ax1.set_yticks([])
+    ax1.scatter(X_train[0, :], X_train[1, :], s=5)
+    ax1.set_xlim(lb, ub)
+    ax1.set_ylim(lb, ub)
+
+    # Plot prediction with gradient enhancement
+    ax2 = fig.add_subplot(spec[0, 1])
+    ax2.contour(X1, X2, Y_pred, 20, cmap='RdGy')
+    anno_opts = dict(xy=(0.5, 1.075), xycoords='axes fraction', va='center', ha='center')
+    ax2.annotate(title, **anno_opts)
+    anno_opts = dict(xy=(0.5, -0.05), xycoords='axes fraction', va='center', ha='center')
+    ax2.annotate('X1', **anno_opts)
+    ax2.set_xticks([])
+    ax2.set_yticks([])
+
+    plt.show()
+
+
+def run_demo(alpha=0.1, beta1=0.9, beta2=0.99, lambd=0.1, gamma=1.0, deep=3, wide=6, batches=32, iterations=30, epochs=100):
```

Comment:
should be moved in test or example file

Reply:
Sounds good. Created a new package under example with appropriate example as recommended.